src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -13,44 +15,94 @@ namespace Microsoft.CodeAnalysis
         {
-            _sourcesAdded.Add(hintName, source);
+            if (string.IsNullOrWhiteSpace(hintName))
+            {
+                throw new ArgumentNullException(nameof(hintName));
+            }
+
+            // allow any identifier character or [.,-_ ()[]{}]
+            for (int i = 0; i < hintName.Length; i++)
+            {
+                char c = hintName[i];
+                if (!UnicodeCharacterUtilities.IsIdentifierPartCharacter(c)
+                    && c != '.'
+                    && c != ','
+                    && c != '-'
+                    && c != '_'
+                    && c != ' '
+                    && c != '('
+                    && c != ')'
+                    && c != '['
+                    && c != ']'
+                    && c != '{'
+                    && c != '}')
+                {
+                    throw new ArgumentException(string.Format(CodeAnalysisResources.HintNameInvalidChar, c, i), nameof(hintName));
+                }
+            }
+
+            hintName = AppendExtensionIfRequired(hintName);
+            if (this.Contains(hintName))
+            {
+                throw new ArgumentException(CodeAnalysisResources.HintNameUniquePerGenerator, nameof(hintName));
+            }
+
+            _sourcesAdded.Add(new GeneratedSourceText(hintName, source));
         }
 
         public void RemoveSource(string hintName)
         {
-            _sourcesAdded.Remove(hintName);
+            hintName = AppendExtensionIfRequired(hintName);
+            for (int i = 0; i < _sourcesAdded.Count; i++)
+            {
+                if (_hintNameComparer.Equals(_sourcesAdded[i].HintName, hintName))
+                {
+                    _sourcesAdded.Remove(_sourcesAdded[i]);
+                    return;
+                }
+            }
+        }
+
+        public bool Contains(string hintName)
+        {
+            hintName = AppendExtensionIfRequired(hintName);
+            for (int i = 0; i < _sourcesAdded.Count; i++)
+            {
+                if (_hintNameComparer.Equals(_sourcesAdded[i].HintName, hintName))
+                {
+                    return true;
+                }
+            }
+            return false;
         }
 
-        public bool Contains(string hintName) => _sourcesAdded.ContainsKey(hintName);
+        internal ImmutableArray<GeneratedSourceText> ToImmutableAndFree() => _sourcesAdded.ToImmutableAndFree();
 
-        internal ImmutableArray<GeneratedSourceText> ToImmutableAndFree()
+        private static string AppendExtensionIfRequired(string hintName)
         {
-            // https://github.com/dotnet/roslyn/issues/42627: This needs to be consistently ordered
-            ArrayBuilder<GeneratedSourceText> builder = ArrayBuilder<GeneratedSourceText>.GetInstance();
-            foreach (var (hintName, sourceText) in _sourcesAdded)
+            if (!Path.GetExtension(hintName).Equals(".cs", StringComparison.OrdinalIgnoreCase))
```

Comment:
Did you consider using `hintName.EndsWith(...)` here? Using `GetExtension` means that operations like `Contains` end up being allocating.